host-controller/src/main/java/org/jboss/as/host/controller/jvm/JvmType.java

```diff
@@ -0,0 +1,193 @@
+     * @param forLaunch {@code true} if the created object will be used for launching servers; {@code false}
+     *                              if it is simply a data holder. A value of {@code true} will disable
+     *                              some validity checks and may disable determining if the JVM is modular
+     * @return the {@code JvmType}. Will not return {@code null}
+     *
+     * @throws IllegalStateException if the {@code JvmType} cannot be determined.
+     */
+    public static JvmType createFromJavaHome(final String javaHome, boolean forLaunch) {
+        if (javaHome == null || javaHome.trim().equals("")) {
+            throw HostControllerLogger.ROOT_LOGGER.invalidJavaHome(javaHome);
+        }
+        final File javaHomeDir = new File(javaHome);
+        if (forLaunch && !javaHomeDir.exists()) {
+            throw HostControllerLogger.ROOT_LOGGER.invalidJavaHome(javaHomeDir.getAbsolutePath());
+        }
+        final File javaBinDir = new File(javaHomeDir, BIN_DIR);
+        if (forLaunch && !javaBinDir.exists()) {
+            throw HostControllerLogger.ROOT_LOGGER.invalidJavaHomeBin(javaBinDir.getAbsolutePath(), javaHomeDir.getAbsolutePath());
+        }
+        final File javaExecutable = new File(javaBinDir, JAVA_EXECUTABLE);
+        if (forLaunch && !javaExecutable.exists()) {
+            throw HostControllerLogger.ROOT_LOGGER.cannotFindJavaExe(javaBinDir.getAbsolutePath());
+        }
+        return new JvmType(forLaunch, isModularJvm(javaExecutable.getAbsolutePath(), forLaunch), javaExecutable.getAbsolutePath());
+    }
+
+    private static boolean isModularJvm(final String javaExecutable, final boolean forLaunch) {
+        if (forLaunch) {
+            try {
+                return 0 == new ProcessBuilder(javaExecutable, "--add-modules=java.se", "-version").start().waitFor();
+            } catch (Throwable t) {
+                throw HostControllerLogger.ROOT_LOGGER.cannotFindJavaExe(javaExecutable);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Create a {@code JvmType} based on the location of the java executable.
+     * @param javaExecutable the location of the java executable. Cannot be {@code null} or empty
+     * @param forLaunch {@code true} if the created object will be used for launching servers; {@code false}
+     *                              if it is simply a data holder. A value of {@code true} will disable
+     *                              some validity checks and may disable determining if the JVM is modular
+     * @return the {@code JvmType}. Will not return {@code null}
+     */
+    public static JvmType createFromJavaExecutable(final String javaExecutable, boolean forLaunch) {
+        assert javaExecutable != null;
+
+        if (javaExecutable.equals(JAVA_WIN_EXECUTABLE) || javaExecutable.equals(JAVA_UNIX_EXECUTABLE)) {
+            return createFromEnvironmentVariable(forLaunch);
```

Comment:
I'm not sure if this is correct. What if the `java` executable lives in the `$PATH`, but the `JAVA_HOME` isn't set or the system property is set to a different JVM? Maybe that's not possible, but if it is just something to consider.